File: snippets/lesson-4-switchmap.cs
using System;
using System.Reactive.Linq;
using System.Threading.Tasks;

IObservable<string> typeStream = Observable.Interval(TimeSpan.FromMilliseconds(200))
    .Select(_ => "User typed something...");

IObservable<string> resultsStream = typeStream
    .Select(term => Observable.FromAsync(async () =>
    {
        await Task.Delay(500);
        return $"Results for: {term}";
    }))
    .Switch();

resultsStream.Subscribe((string data) => Console.WriteLine(data));

File: snippets/lesson-4-switchmap.ts
import { fromEvent, Observable, of } from 'rxjs';
import { switchMap, delay } from 'rxjs/operators';

const searchInput = document.getElementById('search') as HTMLInputElement;
const typeStream$: Observable<Event> = fromEvent(searchInput, 'input');

const results$: Observable<string> = typeStream$.pipe(
  switchMap((event: Event) => {
    const term = (event.target as HTMLInputElement).value;
    return of(`Results for: ${term}`).pipe(delay(500));
  }),
);

results$.subscribe((data: string) => console.log(data));